Clear codes support (MIL)

samifruit514 opened the pull request:
# Description

Support MIL clearing. Please let me know if I'm going the right way

# Checklist

- [ ] Running `go test` locally is successful
- [ ] **VERSION** has been changed
- [ ] Changes has been documented in **CHANGELOG.md**

samifruit514 commented:
It's for Feature 9 (F9)

rzetterberg commented:
Hey, @samifruit514!

I quickly read through your changes and it seems like you have done the following changes: 

- Migrated the project to go modules
- Improved readability by replacing magic numbers by well named constants
- Added a new OBD-command that clears the trouble codes

I think these are great changes. So thanks for that!

The only feedback I have is whether the new command should be a subtype of `UIntCommand` since it doesn't actually receive any data back. I'm wondering what will happen when this command is actually run, when `SetValue` is called on the command, etc. Maybe you could add a test case to [`TestParseOBDResponse`](https://github.com/rzetterberg/elmobd/blob/master/device_test.go#L68)?

samifruit514 commented:
Thanks  for the feedback. Yes this is the purpose of the PR. I should have detailed it more when I opened it.

For the UIntCommand, you are absolutely right. I removed it and also changed the structs/interfaces a bit. I splitted the OBDCommand in 2: normal OBDCommand et OBDCInteractive. The original would be the interactive one. That way, you can run an interactive for the orginal commands (get values & parse) and if you want some commands that doesnt require result parsing, you can use OBDCommand

Let me know what you think.

Thanks

samifruit514 commented:
I tried to split the command into 1 interactive (with output) and 1 non-interactive (no output like the clear troubles code one). The result was too complicated as we could execute clear codes troubles as a interactive command, which didn't make any sense. I reverted the stuff and keep one  ODBCCommand to make things simpler. I added a test case like you requested. Please let me know what you think.

rzetterberg commented on commands.go:
```diff
@@ -852,6 +860,30 @@ func (cmd *Part5Supported) SetValue(result *Result) error {
+	return ""
+}
+
+func NewService04Command(parameterID byte, dataWidth byte, key string) BaseCommand {
```
I don't think hiding one argument justifies the need to introduce a new helper. 

Also it's not consistent with how service 01 commands are created, they don't have a `NewService01Command` helper.

samifruit514 replied:
You're absolutely right. Running a ServiceXXCommand  doesn't make sense (as opposed to running a more concrete command like "EngineRPM".  Sorry for that. But this makes me think about something: Should we change BaseCommand to  baseCommand (not  exported) since it  wouldn't make sense for a user to run a  "raw" base command? It would kind of force users/devs to add  the command if it's missing.

rzetterberg replied:
Yes, great suggestion! 

I think that will make it more straightforward for users of the library how you use the commands and more clear for the developers of the library what role `BaseCommand` has. 

It also aligns with the goal of the library to provide concrete types for all the standard OBD-II PIDs (as defined by SAE J1979). 

If people want to create custom commands that are not in the standard, they can still create their own concrete types that implement the `OBDCommand` interface.

rzetterberg commented on commands_test.go:
```diff
@@ -16,3 +16,16 @@ func TestMonitorStatusResult(t *testing.T) {
+	service01cmd := &Part1Supported{
+		BaseCommand{SERVICE_01_ID, 0, 4, "supported_commands_part1"},
+		UIntCommand{},
+	}
```
How come we are constructing this type manually and not using the constructor `NewPart1Supported?

samifruit514 replied:
good catch. Thanks

rzetterberg commented:
I saw your previous change, but haven't been able to make time to comment on it. 

My comment would be along the lines that I don't want to split `OBDCommand` into 2 separate interfaces and that I want the change to be implemented simpler. 

So I guess you somehow read my mind, good job :laughing: 

Anyway, I've made a review with minor changes, see above :point_up:

jrcichra commented:
Awesome! I just found this project and this was the missing feature I was looking for. I'll be testing this out in my 2009 Sonata.

If I find any issues I'll be sure to open a PR. Thanks for writing a Go version :smile:

samifruit514 commented:
np, it was a pleasure!
